source/common/router/upstream_request.cc

```diff
@@ -111,14 +111,23 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
 
   // The router checks that the connection pool is non-null before creating the upstream request.
   auto upstream_host = conn_pool_->host();
+  absl::optional<Upstream::ClusterInfoConstSharedPtr> cluster_info =
+      parent_.callbacks()->streamInfo().upstreamClusterInfo();
+
   Tracing::HttpTraceContext trace_context(*parent_.downstreamHeaders());
+  Tracing::UpstreamContext upstream_context(upstream_host.get(),           // host_
+                                            &upstream_host->cluster(),     // cluster_
+                                            Tracing::ServiceType::Unknown, // service_type_
+                                            false                          // async_client_span_
```

Comment:
`async_client_span_` can be true here as async client can be the caller . You can try some tests in callout filters like ext_proc

Let's take one step back and rethink about design: 

(1) Do we must have such differentiation in this converging place? Alternatively, can we do it in http async client and gRPC client respectively instead? I don't have too much context here cc @stevenzzzz who has raised https://github.com/envoyproxy/envoy/issues/33167.

(2) If we need (1), then what we are differentiating here are 3 things: (a) envoy-gRPC sidestream (b) http sidestream (c) upstream.  First, we can distinguish between sidestream (a and b) and upstream. Then further identify a and b within sidestream.

(2) is doable based on my previous experience, but let's focus on the first item from design perspective.

Reply:
Thanks for the review, Tianyu! I see your points, and to clarify. 

We are using the trace span `injectContext()` in a tracer and this gives the ability for tracer to know the async client is calling inject context. And that is enough as current stage, even the shared injectContext is sharing. The opposite may be wrong for upstream as you noticed.

Reply:
I am wondering what will happen if this code here is actually triggered by `async_client`?

It seems that you will first set `async_client_span_` to TRUE in http async_client for example, then later here you will set `async_client_span_` to FALSE for async_client path.  Will this override/confuses tracer?

Reply:
Just sync with Boteng offline, in this PR the injection here essentially will be ignored by sidestream as it has already done injection earlier.

The best thing we could do here is to see if we can find a upstream exclusive place in router/upstream to perform such injection for upstream. 

If it is hard to do at this stage or involve some interruptive changes that may not be worthwhile, I am OK with this approach.

cc @wbpcode

Reply:
Thanks Tianyu! Added a TODO and comments to further distinguish it. We can iterate from current stage I think, if we think this is a real issue.

Reply:
Yeah, when async client is used, all spans in `UpstreamRequest` would be null or NullSpan and the injecting would be no-op.